File: csharp/Ice/asyncInvocation/CalculatorI.cs
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// *********************************************************************

using Demo;
using System;

public class CalculatorI : CalculatorDisp_
{
    public override int add(int x, int y, Ice.Current current)
    {
        return x + y;
    }

    public override int subtract(int x, int subtrahend, Ice.Current current)
    {
        return x - subtrahend;
    }

    public override int divide(int numerator, int denominator, out int remainder, Ice.Current current)
    {
        if(denominator == 0)
        {
            throw new Demo.DivideByZeroException();
        }
        remainder = numerator % denominator;
        return numerator / denominator;
    }

    public override int square(int x, Ice.Current current)
    {
        return x * x;
    }

    public override double squareRoot(int x, Ice.Current current)
    {
        if(x < 0)
        {
            throw new NegativeRootException();
        }
        return Math.Sqrt(x);
    }

    public override void shutdown(Ice.Current current)
    {
        current.adapter.getCommunicator().shutdown();
    }
}

File: java/Ice/asyncInvocation/CalculatorI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import com.zeroc.Ice.Current;

import com.zeroc.demos.Ice.asyncInvocation.Demo.*;

public class CalculatorI implements Calculator
{
    @Override
    public int add(int x, int y, Current current)
    {
        return x + y;
    }

    @Override
    public int subtract(int x, int subtrahend, Current current)
    {
        return x - subtrahend;
    }

    @Override
    public DivideResult divide(int numerator, int denominator, Current current) throws DivideByZeroException
    {
        if(denominator == 0)
        {
            throw new DivideByZeroException();
        }
        return new DivideResult(numerator / denominator, numerator % denominator);
    }

    @Override
    public int square(int x, Current current)
    {
        return x * x;
    }

    @Override
    public double squareRoot(int x, Current current) throws NegativeRootException
    {
        if(x < 0)
        {
            throw new NegativeRootException();
        }
        return Math.sqrt(x);
    }

    @Override
    public void shutdown(Current current)
    {
        current.adapter.getCommunicator().shutdown();
    }
}
